### PR TITLE
Video: Add support for selecting or uploading a custom image to serve as the video poster

### PR DESCRIPTION
**Pull request by donnapep**
This PR adds a `VideoEditor` block, which enables choosing a custom VideoPress poster by either selecting a particular frame of the video or by uploading a custom image.

### Testing
You will need to test using a non-Jetpack site with VideoPress enabled until support is added for Jetpack sites as well.

The Video Editor can be accessed in the editor by selecting "Add Media" from the toolbar dropdown, selecting a video, clicking the "Edit" button, and then clicking the "Edit Thumbnail" button:

screen shot 2017-02-15 at 11 42 15 am

screen shot 2017-02-15 at 11 37 29 am

If the video cannot be loaded or an error occurs when updating the poster, the following message is shown:

screen shot 2017-02-15 at 11 38 09 am

This PR is the master but has been split out into separate PRs (#11633, #11637, #11648, #11653)

### CODE REVIEW

File: client/blocks/video-editor/video-editor-buttons.jsx [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+					className="button video-editor__buttons-button"
+					onUploadImage={ onUploadImage }>
+					{ translate( 'Upload Image' ) }
+				</UploadButton>
+				<Button
+					className="video-editor__buttons-button"
+					primary
+					onClick={ onSelectFrame }
+				>
+					{ translate( 'Select Frame' ) }
```
Comment (a8ci18n) [inline client/blocks/video-editor/video-editor-buttons.jsx]: Hi! I've found a possible matching string that has already been translated **28** times:
`translate( 'Frames' )` *ES Score: 7.57*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/post-editor/media-modal/detail/detail-item.jsx [hunk 2]
```diff
@@ -57,17 +57,23 @@ class EditorMediaModalDetailItem extends Component {
...
 
-		if ( 'image' !== mimePrefix ) {
+		if ( ! includes( [ 'image', 'video' ], mimePrefix ) ) {
 			return null;
 		}
 
+		let editText = translate( 'Edit Image' );
+
+		if ( 'video' === mimePrefix ) {
+			editText = translate( 'Edit Thumbnail' );
```
Comment (a8ci18n) [inline client/post-editor/media-modal/detail/detail-item.jsx]: Hi! I've found a possible matching string that has already been translated **79** times:
`translate( 'Thumbnail' )` *ES Score: 9.74*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/video-thumbnail-picker)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/blocks/video-editor/index.jsx [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+		const {
+			onCancel,
+			translate,
+		} = this.props;
+
+		return (
+			<Notice
+				status="is-error"
+				showDismiss={ true }
+				text={ translate( 'We are unable to edit this video.' ) }
```
Comment (a8ci18n) [inline client/blocks/video-editor/index.jsx]: Hi! I've found a possible matching string that has already been translated **18** times:
`translate( 'We are unable to edit this image.' )` *ES Score: 8.31*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/blocks/video-editor/video-editor-buttons.jsx [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+					className="button video-editor__buttons-button"
+					onUploadImage={ onUploadImage }>
+					{ translate( 'Upload Image' ) }
+				</UploadButton>
+				<Button
+					className="video-editor__buttons-button"
+					disabled={ ! isVideoLoaded }
+					onClick={ onSelectFrame }
+					primary>
+					{ translate( 'Select Frame' ) }
```
Comment (a8ci18n) [inline client/blocks/video-editor/video-editor-buttons.jsx]: Hi! I've found a possible matching string that has already been translated **28** times:
`translate( 'Frames' )` *ES Score: 7.80*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/blocks/video-editor/video-editor-buttons.jsx [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+					onClick={ onUploadImageClick }
+					onUploadImage={ onUploadImage }>
+					{ translate( 'Upload Image' ) }
+				</UploadButton>
+				<Button
+					className="video-editor__buttons-button"
+					disabled={ ! isVideoLoaded || isPosterUpdating }
+					onClick={ onSelectFrame }
+					primary>
+					{ translate( 'Select Frame' ) }
```
Comment (a8ci18n) [inline client/blocks/video-editor/video-editor-buttons.jsx]: Hi! I've found a possible matching string that has already been translated **28** times:
`translate( 'Frames' )` *ES Score: 8*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/blocks/video-editor/index.jsx [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
```
Comment (a8ci18n) [inline client/blocks/video-editor/index.jsx]: Hi! I've found a possible matching string that has already been translated **18** times:
`translate( 'We are unable to edit this image.' )` *ES Score: 8*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

Comment (a8ci18n) [inline client/blocks/video-editor/index.jsx]: Hi! I've found a possible matching string that has already been translated **18** times:
`translate( 'We are unable to edit this image.' )` *ES Score: 8*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

Comment (a8ci18n) [inline client/blocks/video-editor/index.jsx]: Hi! I've found a possible matching string that has already been translated **18** times:
`translate( 'We are unable to edit this image.' )` *ES Score: 8*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

Comment (a8ci18n) [inline client/blocks/video-editor/index.jsx]: Hi! I've found a possible matching string that has already been translated **18** times:
`translate( 'We are unable to edit this image.' )` *ES Score: 8*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

Comment (a8ci18n) [inline client/blocks/video-editor/index.jsx]: Hi! I've found a possible matching string that has already been translated **18** times:
`translate( 'We are unable to edit this image.' )` *ES Score: 8*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by beaulebens** [comment]
Note: looks like this PR needs rebasing/conflict resolution.

I tried going through the Media menu item (in the sidebar, not enabled in production yet), and clicking to edit a video. When I clicked "Edit Thumbnail" from that view, nothing happens. No errors or anything, just nothing happens.

When I went in via a Post (Post > New, Add Media, Edit Video, Edit Thumbnail), I saw the editor!

I tried clicking to [Select Frame], but got this error: https://cloudup.com/cJ2tawIC_Xy

Doing it a few more times, it worked in other cases, so I'm not sure why that one didn't.

I did notice that I ended up with some 404s in my console (for missing thumbnail images, apparently). It seemed to generate a thumbnail image with a typo in the filename, don't know if that's related or not. The filename I got was `img_2557_scruberthumbnail_0.jpg` (scruber).

Other than that, it seemed to work quite nicely as far as functionality goes!

The UI feels a little clunky to be honest, and a little jarring. I'm not sure what a better option would be, without a significant overhaul, so for now, at least this works, and allows users to both upload their own, and pick from frames, so it's a significant improvement.

File: client/blocks/video-editor/video-editor-buttons.jsx [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+				onClick={ onUploadImageClick }
+				onUploadImage={ onUploadImage }>
+				{ translate( 'Upload Image' ) }
+			</UploadButton>
+			<Button
+				className="video-editor__buttons-button"
+				disabled={ ! isVideoLoaded || isPosterUpdating }
+				onClick={ onSelectFrame }
+				primary>
+				{ translate( 'Select Frame' ) }
```
Comment (a8ci18n) [inline client/blocks/video-editor/video-editor-buttons.jsx]: Hi! I've found a possible matching string that has already been translated **28** times:
`translate( 'Frames' )` *ES Score: 8*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by beaulebens** [comment]
I got the same error (unable to edit this video) when trying again, and after getting that, if I try to capture a different frame, nothing seems to happen. When I checked in my console, I had:

screen shot 2017-03-06 at 2 07 51 pm

It works in the Media section now, huzzah!

After trying it a few more times, it seems that most of the time, the _first_ attempt on any of my videos fails, but then if I exit out of the thumbnail-picker UI, and try again, it works. I tried on about 5 different videos and only one worked first time, but they all worked after the first try.

I did also notice a number of `404` errors where the thumbnails I had set didn't seem to load/weren't available yet. Even refreshing the page didn't seem to bring them up though, so I don't know if there was an error in the thumbnail generation side of things, or what happened there.

**Comment by donnapep** [comment]
@beaulebens Thanks for the review(s)! Would you be able to grab a screenshot of the DevTools Network tab, filtered on "poster", when you see that error? The "We are unable to edit this video." message shows when an error response is received back from the server. In the meantime, I'll try to reproduce those issues on my end. Cheers!

**Comment by donnapep** [comment]
@beaulebens Ah! So I can reproduce this when uploading a new video and trying to select a frame before VideoPress has finished processing the video. Can you confirm that is the same scenario in which you see the error message? Thx!

**Comment by beaulebens** [comment]
I just tested again, and got the error again. Note that this is _not_ a new video; it's actually been uploaded since August 2015 (maybe being really old has something to do with the problem?).

Here is the error reported from the `/videos/{hash}/poster` endpoint (which had an HTTP status of `200`):

`{"code":404,"headers":[{"name":"Content-Type","value":"application\/json"}],"body":{"error":"thumbnail_error","message":"Could not download the thumbnail"}}`

Since you mentioned it, I tried uploading a new video and immediately editing the thumbnail as well.
- drag/dropped video into the media section
- as soon as it finished uploading, clicked the edit/pencil icon
- clicked edit thumbnail
- got a couple of 404s, but then pretty quickly saw a preview and was able to select a new one

Right after that, I tried selecting another one (on the video I had just uploaded) and then got the same error as above!

**Comment by dbtlr** [comment]
@beaulebens This is an error that happens when trying to return a thumbnail from the VideoPress thumbnail server. It essentially tries to download this URL: `https://thumbs.videopress.com/{guid}?s={seconds}&c=1` and set it as the thumbnails. The downloading step is failing though and the given error is being emitted.

I'll take a look into the thumbnail server, which until now has been a fairly black box for me, and see if there is anyway to track down specific errors we might be seeing.

**Comment by dbtlr** [comment]
@donnapep I'm working with the error that @beaulebens is finding, but it is not related at all to your Calypso code, but is instead related to how the thumbnail server is streaming the video file in for processing. This only seems to be happening with MOV files, so I don't think it is systemic enough to be a blocker. Plus it will resolve itself soon, as we figure out a way to solve it. :)

**Comment by dbtlr** [comment]
Just a note @donnapep: The error found here with those particular thumbnails should be resolved on the server.

**Comment by donnapep** [comment]
This was the master PR for video editor, but it has since been separated out into multiple PRs.

**Pull request closed without merging**